tests/test_gain.cpp

```diff
@@ -1,90 +1,107 @@
-
- THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
-WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ Permission is hereby granted, free of charge, to any person obtaining a copy of
+this software and associated documentation files (the “Software”), to deal in
+the Software without restriction, including without limitation the rights to
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
 
 #include <cstddef>
 #include <math.h>
 
 #include <gtest/gtest.h>
+#include <type_traits>
 
-#include "src/core/pooya.hpp"
 #include "src/core/helper.hpp"
+#include "src/core/pooya.hpp"
 #include "src/core/solver.hpp"
 #include "src/misc/gp-ios.hpp"
 
-class TestGain : public testing::Test
-{
-public:
-    TestGain()
-    {
-        //
-    }
+template <typename T> class TestScalarGain : public testing::Test {
+protected:
+  void SetUp() override {}
 };
 
-TEST_F(TestGain, ScalarGain)
-{
-    // test parameters
-    const double x = 3.7;
-
-    // model setup
-    pooya::Model model;
-    pooya::Gain gain("gain", 2.0);
-    pooya::ScalarSignalId s_x = model.create_scalar_signal("x");
-    pooya::ScalarSignalId s_y = model.create_scalar_signal("y");
-    model.add_block(gain, s_x, s_y);
-
-    // simulator setup
-    pooya::Simulator sim(model,
-        [&](pooya::Model&, double /*t*/) -> void
-        {
-            s_x->set(x);
-        });
-
-    // do one step
-    sim.init(0.0);
-
-    // verify the results
-    EXPECT_NEAR(gain.gain() * x, *s_y, 1e-10);
+typedef ::testing::Types<int, float, double> MyTypes;
+TYPED_TEST_SUITE(TestScalarGain, MyTypes);
+
+TYPED_TEST(TestScalarGain, checkScalarGain) {
+  using Type = TypeParam;
+
+  // model setup
+  pooya::Model model;
+  pooya::Gain gain("gain", 2.0);
+  pooya::ScalarSignalId s_x = model.create_scalar_signal("x");
+  pooya::ScalarSignalId s_y = model.create_scalar_signal("y");
```

Comment:
Why not make these type-specific? Have you tried `pooya::GainT<Type>`? The same applies to the signal IDs too.

FYI: Soon I will open a PR to make major changes to `pooya`. These changes will make its interface simpler and faster. In brief:
- The `Model` class will become obsolete
- Signals will become standalone objects
- New signal wrappers will be introduced. These wrappers will make it more intuitive to get/set signal values.

Reply:
When do you push the updates to the repo? We could merge this PR and then revise the tests once the update is there.

Reply:
You are right @modjtabaf . I forgot about types in signals and gain block. On hindsight, I feel making a parameterized test was not a good idea. It introduced redundant code to handle different signal types and gain types. So, I circled back to make a separate test for `int`. This seems to be a  pragmatic solution.

Reply:
I am pushing the changes on an ongoing basis. It is in a stable state now. Once the merge conflict is resolved, this should be ready to go.